UCBChangesToMorphic.js

```diff
@@ -0,0 +1,296 @@
+HandMorph.prototype.processDrop = function (event) {
+/*
+    find out whether an external image or audio file was dropped
+    onto the world canvas, turn it into an offscreen canvas or audio
+    element and dispatch the
+
+        droppedImage(canvas, name)
+        droppedSVG(image, name)
+        droppedAudio(audio, name)
+        droppedText(text, name, type)
+
+    events to interested Morphs at the mouse pointer.
+
+    In case multiple files are dropped simulateneously also displatch
+    the events
+
+        beginBulkDrop()
+        endBulkDrop()
+
+    to Morphs interested in bracketing the bulk operation
+*/
+    var files = event instanceof FileList ? event
+                : event.target.files || event.dataTransfer.files,
+        file,
+        fileCount,
+        url = event.dataTransfer ?
+                event.dataTransfer.getData('URL') : null,
+        txt = event.dataTransfer ?
+                event.dataTransfer.getData('Text/HTML') : null,
+        suffix,
+        src,
+        target = this.morphAtPointer(),
+        img = new Image(),
+        canvas,
+        i;
+
+    function readSVG(aFile) {
+        var pic = new Image(),
+            frd = new FileReader(),
+            trg = target;
+        while (!trg.droppedSVG) {
+            trg = trg.parent;
+        }
+        pic.onload = () => {
+            trg.droppedSVG(pic, aFile.name);
+            bulkDrop();
+        };
+        frd = new FileReader();
+        frd.onloadend = (e) => pic.src = e.target.result;
+        frd.readAsDataURL(aFile);
+    }
+
+    function readImage(aFile) {
+        var pic = new Image(),
+            frd = new FileReader(),
+            trg = target;
+        while (!trg.droppedImage) {
+            trg = trg.parent;
+        }
+        
+        /* UCB Script Pic Edit:
+            *  - Added new process for accepting images dropped on script window
+            *  - Images are searched for XML metadata, and if present, users are able to:
+            *      1. Process the script as an XML file and load script
+            *      2. Process the script as an image and load it into the costumes
+            *  - Base code credit given do Snap-Forum user Dardoro
+            */
+        function arr2Str(arr) { 
+            return arr.reduce((acc, b) => acc + String.fromCharCode(b), "");
+        };
+        
+        let ide = world.children[0],
+            url = event.dataTransfer?.getData( "text/uri-list"),
+            file = event.dataTransfer?.files?.[0],
+            menu = new MenuMorph(this, "Process Dropped Object");
+
+        pic.onload = () => {
+            canvas = newCanvas(new Point(pic.width, pic.height), true);
+            canvas.getContext('2d').drawImage(pic, 0, 0);
+
+            (async () => {
+                if (!file && url) {
+                    file = await fetch("https://api.allorigins.win/raw?url="+url)
```

Comment:
I'm little nervous about hard-coding the url of someone else's server into the function that handles loading graphics. What if the server is down? Would it make sense to add a fallback option in that case, or - forgive my ignorance - is that already handled by the async nature of this?

Reply:
AFAIK, this is a way of getting around CORS limitations. I would think it's OK to just remove this and raise an error if a file cannot be read.

Reply:
It was primarily intended for drag&drop images from Discourse posts, served by CDN. So depending on the expected use cases may be not necessary (or CORS headers can be sent to allow snap's domains).